docs/architecture.md

```diff
@@ -0,0 +1,185 @@
+Starting with Pydantic V2, part of the codebase is written in Rust in a separate package called `pydantic-core`.
+This architecture documentation will partly focus on how the two `pydantic` and `pydantic-core` packages interacts together.
+
+Usage of the Pydantic library can be divided into two parts:
+
+- Model definition, done in the `pydantic` package.
+- Model validation and serialization, done in the `pydantic-core` package.
+
+## Model definition
+
+Whenever a Pydantic [`BaseModel`][pydantic.main.BaseModel] is defined, the metaclass
+will analyze the body of the model to collect a number of elements:
+
+- Defined annotations to build model fields (collected in the [`model_fields`][pydantic.main.BaseModel.model_fields] attribute).
+- Model configuration, set with [`model_config`][pydantic.main.BaseModel.model_config].
+- Additional validators/serializers.
+- Private attributes, identification of generic parametrization, etc.
+
+### Communicating between `pydantic` and `pydantic-core`
+
+Once the model class is successfully built, a core schema for the model will be constructed
+and set as the [`__pydantic_core_schema__`][pydantic.main.BaseModel.__pydantic_core_schema__]
+attribute (this is done by the `GenerateSchema` class). A core schema is a structured Python
+dictionary (represented using [`TypedDict`][typing.TypedDict] definitions) representing a
+specific validation and serialization logic. It is the core data structure used to communicate
+between the `pydantic` and `pydantic-core` packages. Every core schema has a required `type` key,
+and extra properties depending on this `type`.
+
+!!! note
+    It is not possible to define a custom core schema. A core schema needs to be understood by the
+    `pydantic-core` package, and as such we only support a fixed number of core schema types.
+    This is also part of the reason why the `GenerateSchema` is private and not documented.
```

Comment:
Unfortunately (imo) we do sort of document it, though we've sort of shied away from further documentation bc we want to reserve some rights to modify how we generate core schema. So we might want to reword here.